lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1521 @@
+//===- InferDomains.cpp - Infer and Check FIRRTL Domains ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// InferDomains implements FIRRTL domain inference and checking. This pass is a
+// bottom-up transform acting on modules. For each module, we ensure there are
+// no domain crossings, and we make explicit the domain associations of ports.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/Debug.h"
+#include "circt/Support/Namespace.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-infer-domains"
+
+namespace circt {
+namespace firrtl {
+#define GEN_PASS_DEF_INFERDOMAINS
+#include "circt/Dialect/FIRRTL/Passes.h.inc"
+} // namespace firrtl
+} // namespace circt
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+
+//====--------------------------------------------------------------------------
+// Helpers.
+//====--------------------------------------------------------------------------
+
+using PortInsertions = SmallVector<std::pair<unsigned, PortInfo>>;
+
+template <typename T>
+bool shouldInfer(T op, InferDomainsMode mode) {
+  return op.isPublic() ? shouldInferPublicModules(mode)
+                       : shouldInferPrivateModules(mode);
+}
+
+/// From a domain info attribute, get the domain-type of a domain value at
+/// index i.
+StringAttr getDomainPortTypeName(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return nullptr;
+  auto ref = cast<FlatSymbolRefAttr>(info[i]);
+  return ref.getAttr();
+}
+
+/// From a domain info attribute, get the row of associated domains for a
+/// hardware value at index i.
+auto getPortDomainAssociation(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return info.getAsRange<IntegerAttr>();
+  return cast<ArrayAttr>(info[i]).getAsRange<IntegerAttr>();
+}
+
+/// Return true if the value is a port on the module.
+bool isPort(BlockArgument arg) {
+  return isa<FModuleOp>(arg.getOwner()->getParentOp());
+}
+
+/// Return true if the value is a port on the module.
+bool isPort(Value value) {
+  auto arg = dyn_cast<BlockArgument>(value);
+  if (!arg)
+    return false;
+  return isPort(arg);
+}
+
+/// Returns true if the value is driven by a connect op.
```

Comment:
Possibly clarify this is only for Domain-typed values? `TypedValue<DomainType>` is a thing if you want.

Mostly was wondering whether this meant /directly/ driven (which is what's implemented) or indirectly by driving through indexing operations.